File: finder4j-backtrack/src/test/java/com/github/k0zka/finder4j/backtrack/termination/OrTerminationStrategyTest.java
package com.github.k0zka.finder4j.backtrack.termination;

import java.util.Arrays;

import org.junit.Assert;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.Mockito;
import org.mockito.runners.MockitoJUnitRunner;

import com.github.k0zka.finder4j.backtrack.State;
import com.github.k0zka.finder4j.backtrack.TerminationStrategy;

@RunWith(MockitoJUnitRunner.class)
public class OrTerminationStrategyTest {
	
	@Mock
	TerminationStrategy<State> strategy1;

	@Mock
	TerminationStrategy<State> strategy2;

	@Mock
	TerminationStrategy<State> strategy3;

	@Mock
	State state;

	@Test
	public void stopWithTrue() {
		Mockito.when(strategy1.stop(Mockito.eq(state))).thenReturn(false);
		Mockito.when(strategy2.stop(Mockito.eq(state))).thenReturn(true);
		Assert.assertTrue(new OrTerminationStrategy<State>(Arrays.asList(strategy1, strategy2, strategy3)).stop(state));
		Mockito.verify(strategy3, Mockito.never()).stop(state);
		Mockito.verify(strategy2, Mockito.times(1)).stop(state);
		Mockito.verify(strategy1, Mockito.times(1)).stop(state);
	}

	@Test
	public void stopWithFalse() {
		Mockito.when(strategy1.stop(Mockito.eq(state))).thenReturn(false);
		Mockito.when(strategy2.stop(Mockito.eq(state))).thenReturn(false);
		Mockito.when(strategy3.stop(Mockito.eq(state))).thenReturn(false);
		Assert.assertFalse(new OrTerminationStrategy<State>(Arrays.asList(strategy1, strategy2, strategy3)).stop(state));
		Mockito.verify(strategy3, Mockito.times(1)).stop(state);
		Mockito.verify(strategy2, Mockito.times(1)).stop(state);
		Mockito.verify(strategy1, Mockito.times(1)).stop(state);
	}

}

File: finder4j-backtrack/src/test/kotlin/com/github/k0zka/finder4j/backtrack/termination/OrTerminationStrategyTest.kt
package com.github.k0zka.finder4j.backtrack.termination

import com.github.k0zka.finder4j.backtrack.State
import com.github.k0zka.finder4j.backtrack.TerminationStrategy
import com.nhaarman.mockito_kotlin.eq
import com.nhaarman.mockito_kotlin.mock
import com.nhaarman.mockito_kotlin.never
import com.nhaarman.mockito_kotlin.times
import com.nhaarman.mockito_kotlin.verify
import com.nhaarman.mockito_kotlin.whenever
import org.junit.Assert
import org.junit.Assert.assertFalse
import org.junit.Test

class OrTerminationStrategyTest {

	private val strategy1: TerminationStrategy<State> = mock()

	private val strategy2: TerminationStrategy<State> = mock()

	private val strategy3: TerminationStrategy<State> = mock()

	private val state: State = mock()

	@Test
	fun stopWithTrue() {
		whenever(strategy1.stop(eq(state))).thenReturn(false)
		whenever(strategy2.stop(eq(state))).thenReturn(true)
		Assert.assertTrue(
				OrTerminationStrategy(listOf(strategy1, strategy2, strategy3)).stop(
						state))
		verify(strategy3, never()).stop(state)
		verify(strategy2, times(1)).stop(state)
		verify(strategy1, times(1)).stop(state)
	}

	@Test
	fun stopWithFalse() {
		whenever(strategy1.stop(eq(state))).thenReturn(false)
		whenever(strategy2.stop(eq(state))).thenReturn(false)
		whenever(strategy3.stop(eq(state))).thenReturn(false)
		assertFalse(
				OrTerminationStrategy(listOf(strategy1, strategy2, strategy3)).stop(
						state))
		verify(strategy3, times(1)).stop(state)
		verify(strategy2, times(1)).stop(state)
		verify(strategy1, times(1)).stop(state)
	}

}
